aten/src/ATen/native/mps/operations/Quantized.mm

```diff
@@ -204,6 +204,458 @@ kernel void int4pack_mm(
+
+template <> struct BlockType<float> {
+  using simdgroup_type8x8 = simdgroup_float8x8;
+  using type4 = float4;
+};
+
+template <> struct BlockType<half> {
+  using simdgroup_type8x8 = simdgroup_half8x8;
+  using type4 = half4;
+};
+#if __METAL_VERSION__ >= 310
+template <> struct BlockType<bfloat> {
+  using simdgroup_type8x8 = simdgroup_bfloat8x8;
+  using type4 = bfloat4;
+};
+#endif
+
+template<typename T>
+float2 get_scale_zero(constant T * scalesAndZeros, uint2 index) {
+    return float2(1.0, 0.0);
+}
+
+template<typename T>
+float2 get_scale_zero_q8(constant T * scalesAndZeros, uint2 index) {
+    T scale = scalesAndZeros[index[0]];
+    return float2(scale, 0.0);
+}
+
+#define BLOCK_SIZE_M 32 // each block takes 32 rows in matrix A
+#define BLOCK_SIZE_N 64 // each block takes 64 rows in matrix B
+#define BLOCK_SIZE_K 32
+#define THREAD_MAT_M 2 // in data loading stage, each thread load 2 simdgroup matrices from matrix A
+#define THREAD_MAT_N 4 // in data loading stage, each thread load 4 simdgroup matrices from matrix B
+#define THREAD_PER_ROW_A 4 // 4 thread for each row in matrix A to load numbers
+#define THREAD_PER_ROW_B 2 // 2 thread for each row in matrix B to load numbers
+#define SG_MAT_SIZE 64 // simdgroup matrix is of shape 8x8
+#define SG_MAT_ROW 8
+
+// T: input type, W: weight type
+template<typename T, typename W, float2 (*get_scale_zero_func)(constant T *, uint2)>
+kernel void kernel_mul_mm(
+    constant T                 * A              [[buffer(0)]],
+    constant char              * B              [[buffer(1)]],
+    constant T                 * scalesAndZeros [[buffer(2)]],
+    device T                   * outputData     [[buffer(3)]],
+    constant uint3             & sizes          [[buffer(4)]],
+    threadgroup char           * shared_memory  [[threadgroup(0)]], // threadgroup buffer at index 0
+    uint3                        tgpig          [[threadgroup_position_in_grid]], // 3d coordinates
+    uint                         tiitg          [[thread_index_in_threadgroup]], // 128 per threadgroup
+    uint                         sgitg          [[simdgroup_index_in_threadgroup]]) {
+
+    using T4 = typename BlockType<T>::type4;
+    using Tsimd8x8 = typename BlockType<T>::simdgroup_type8x8;
+    // sizes: x = M, y = K, z = N
+    // pytorch: M x K @ N x K -> M x N
+    // ggml: K x N @ K x M -> N x M
+    uint32_t M = sizes.x; // M
+    uint32_t K = sizes.y; // K
+    uint32_t N = sizes.z; // N
+    uint32_t nbytes_B = sizeof(W); // number of bytes for one element in B
+    uint32_t nbytes_B_row = nbytes_B * K; // number of bytes for one row in B
+    uint32_t nbytes_A = sizeof(T); // number of bytes for one element in A
+    uint32_t nbytes_A_row = nbytes_A * K; // number of bytes for one row in A
+
+    // shared memory for A and B
+    threadgroup T    * shared_memory_A = (threadgroup T    *)(shared_memory);
+    threadgroup half * shared_memory_B = (threadgroup half *)(shared_memory + 8192);
```

Comment:
Two remarks: - all offsets should be expressed as number of elements multiplied by sizeof elemen
Let's keep A and B types the same
```suggestion
    threadgroup T * shared_memory_B = (threadgroup half *)(shared_memory + 4096 * sizeof(T));
```

Reply:
Added comment, as using `half` here is good for perf (8% improvement) comparing with `bfloat`.